src/jsonrpc.js

```diff
@@ -1,30 +1,104 @@
-   * @param {*} addresses 
-   * @param {*} fromBlock 
-   * @param {*} toBlock 
-   * @param {*} callback 
+   * @param Array addresses 
+   * @param int currentBlock 
+   * @param int toBlock 
+   * @param function callback 
    */
-  constructor(addresses, fromBlock, toBlock, web3, callback = null) {
+  constructor(addresses, currentBlock, toBlock, web3, callback = null) {
     this.addresses = addresses.map(address => address.toLowerCase());
-    this.fromBlock = fromBlock;
+    this.currentBlock = currentBlock;
     this.toBlock = toBlock;
     this.web3Instance = web3.eth;
     this.callback = callback;
+    if (!callback) {
+      logger.info('Warning!: No callback function defined');
+    }
+  }
+
+  /**
+   * This will return the final data structure 
+   * for the transaction resopnse
+   * @param string tx 
+   * @param Object receipt
+   * @param Array logs
+   * @return object
+   */
+  static getTransactionFormat(transaction, receipt, logs) {
+    const receiptResult = receipt;
+
+    receiptResult.logs = logs;
+    return { ...transaction, ...receiptResult };
+  }
+
+  /**
+   * Async function that gets the transaction and transaction receipt 
+   * then get the logs out of the receipt transaction then execute the callback function  
+   * @param string txn
+   */
+  async _scanTransaction(txn) {
+    const txnReceipts = await bluebird.promisify(this.web3Instance.getTransactionReceipt)(txn.hash);
+    const logs = txnReceipts.logs ? txnReceipts.logs.filter(log => isInArray(this.addresses,
+      log.address)) : [];
+
+    if (txn.from && isInArray(this.addresses, txn.from.toLowerCase())) {
+      throw new Error('The address you entered is not a smart contract');
+    }
+
+    // If the smart contract recieved transaction or there's logs execute the callback function 
+    if ((txn.to && isInArray(this.addresses, txn.to.toLowerCase())) || logs.length > 0) {
+      const transactionResult = JsonRpc.getTransactionFormat(txn, txnReceipts, logs);
+      if (this.callback) { await this.callback(transactionResult); }
+    }
+  }
+
+  /**
+   * This function handles the transactions that exist in one block
+   *  and puts them into an array of promises, then executes them.
+   * @param Object block 
+   */
+  async _scanBlockCallback(block) {
+    if (block && block.transactions && Array.isArray(block.transactions)) {
+      const promises = [];
+      for (let i = 0; i < block.transactions.length; i += 1) {
+        const txn = block.transactions[i];
+        promises.push(this._scanTransaction(txn));
+      }
+      await Promise.all(promises);
+    }
   }
 
+
   /**
-   * Scanning all the blocks fromBlock and toBlock 
-   * which are internal variables passed via the constructor
+   * The main function that run scan all the blocks.
    */
   async scanBlocks() {
-    if (this.fromBlock >= this.toBlock) {
-      logger.debug(`Last block number is ${this.fromBlock}`);
+    if (this.currentBlock >= this.toBlock) {
```

Comment:
This check is redundant. if we put it in the beginning of the class as a condition